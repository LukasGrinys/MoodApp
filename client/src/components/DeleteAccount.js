import React, { Component } from 'react';
import axios from 'axios';
import { connect } from 'react-redux';
import FormWrapper from './../widgets/FormWrapper';
import Loading from './Loading/loading';
import ButtonWid from './../widgets/Button';
import MessageBox from './../widgets/MessageBox';
import BackButton from './../widgets/backButton';

import styles from './settings.module.css';

import { deleteUser } from './../actions';


class DeleteAccount extends Component {
    state = {
        password: '',
        password_confirm: '',
        isDisabled: false,
        error: null,
        errorMsg : ''
    }

    handlePassword = (event) => {
        let password = event.target.value;
        this.setState({password, error: null})
    }
    handlePasswordConfirm = (event) => {
        let password_confirm = event.target.value;
        this.setState({password_confirm, error : null})
    }

    renderErrorMessage = () => {
        if (this.state.error === true) {
            return <MessageBox success={!this.state.error} text={this.state.errorMsg}/>
        } else {
            return null
        }
    }

    handleDeleteAccount = () => {
        // Empty fields
        if (this.state.password.trim().length === 0 || this.state.password_confirm.trim().length === 0) {
            this.setState({
                error: true,
                errorMsg : "Fields cannot be empty"
            });
            return;
        }
        // Passwords do not match
        if (this.state.password.trim() !== this.state.password_confirm.trim()) {
            this.setState({
                error: true,
                errorMsg : "Passwords do not match"
            });
            return;
        }
        // 
        this.setState({isDisabled : true});
        let password = this.state.password.trim();
        let userId = this.props.user.data.id;
        const body = {
            password,
            userId
        }
        axios.post(`/api/deleteAccount`, body)
        .then( (response) => {
            if (response.data.error === true) {
                this.setState({
                    error: true,
                    errorMsg : response.data.message,
                    isDisabled : false
                })
            } else {
                this.props.dispatch(deleteUser());
                this.props.history.push('/');
            }
        })
    }

    saveNewPassword = () => {
        if (this.state.isDisabled) return;
        const oldpassword = this.state.currentPassword;
        const newpassword = this.state.newPassword;
        const _id = this.props.user.data.id;
        const body = {
            _id, oldpassword, newpassword
        };
        if (newpassword.trim().length < 6) {
            this.setState({
                passwordChanged: false,
                passwordChangedMessage: "The password is too short. You must pick a password longer than 6 characters"
            })
            return;
        } else {
            this.setState({isDisabled: true})
            axios.post('/api/changePassword', body)
            .then( ({data}) => {
                if (data.error === true) {
                    this.setState({
                        passwordChanged: false,
                        passwordChangedMessage: data.message,
                        isDisabled: false
                    })
                };
                if (data.error === false) {
                    this.setState({
                        passwordChanged: true,
                        passwordChangedMessage: data.message,
                        currentPassword: '',
                        newPassword: '',
                        isDisabled: false
                    });
                }
                if (!data) {
                    this.setState({
                        passwordChanged: false,
                        passwordChangedMessage: "An error occured. Try again later",
                        isDisabled: false
                    })
                }
            })
        }
    }

    render() {
        if (this.props.user) {
            return (
                <div>
                    <BackButton/>
                    <FormWrapper>
                        <h2>Delete account:</h2>
                        <label htmlFor="password">Enter your password:</label>
                        <input type="password" onChange={this.handlePassword}></input>
                        <label htmlFor="password_confirm">Confirm your password:</label>
                        <input type="password" onChange={this.handlePasswordConfirm}></input>
                        {this.renderErrorMessage()}
                        <div className={styles.center}>
                            <div onClick={this.handleDeleteAccount}>
                                <ButtonWid background="#3366FF" color="#FFFFFF" disabled={this.state.isDisabled}>
                                    Delete account
                                </ButtonWid>
                            </div>
                        </div>
                    </FormWrapper>
                </div>
            )
        } else {
            return <Loading/>
        }
    }
}

function mapStateToProps(state) {
    return {
        logs: state.logs
    }
}

export default connect(mapStateToProps)(DeleteAccount);