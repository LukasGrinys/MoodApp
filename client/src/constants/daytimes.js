export const daytimes = {
    morning: 'Morning',
    afternoon: 'Afternoon',
    evening: 'Evening',
    night: 'Night'
}