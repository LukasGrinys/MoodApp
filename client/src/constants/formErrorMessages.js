export const formErrorMessages = {
    required : 'Required field',
    email: 'Enter a proper e-mail address'
}