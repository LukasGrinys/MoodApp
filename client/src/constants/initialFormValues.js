export const initialFormValues = {
    email: '',
    password: '',
    firstName: '',
    lastName: '',
    confirmPassword: '',
    currentPassword: ''
}