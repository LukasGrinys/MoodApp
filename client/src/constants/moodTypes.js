const moodTypes = [
    {
        above: 7.5,
        below: 10,
        text: 'Great'
    },
    {
        above: 5,
        below: 7.5,
        text: 'Good'
    },
    {
        above: 3,
        below: 5,
        text: 'Average'
    },
    {
        above: 0,
        below: 3,
        text: 'Bad'
    }
];

export default moodTypes;