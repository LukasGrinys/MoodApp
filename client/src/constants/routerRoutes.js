export const routerRoutes = {
    home: '/',
    login: '/login',
    logout: '/logout',
    dashboard: '/dashboard',
    addNewLog: '/addlog',
    register: '/register',
    settings: '/settings'
}