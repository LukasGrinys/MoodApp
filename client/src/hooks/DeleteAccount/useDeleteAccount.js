import { useState, useEffect } from 'react';
import { useFormik } from 'formik';
import { initialFormValues } from '../../constants/initialFormValues';
import * as yup from 'yup';
import { formErrorMessages } from '../../constants/formErrorMessages';
import axios from 'axios';
import { useSelector, useDispatch } from 'react-redux';
import { useHistory } from 'react-router-dom';
import { clearUserData } from '../../actions/user/asyncActions';
import { routerRoutes } from '../../constants/routerRoutes';
import Cookie from 'universal-cookie';
const cookie = new Cookie();

export const useDeleteAccount = () => {
    const [deleteAccountStatus, setDeleteAccountStatus] = useState({
        isDeleting: false,
        error: null,
        success: null
    });
    const { isDeleting, success, error } = deleteAccountStatus;
    const userId = useSelector( ({user}) => user.userData.id);
    const history = useHistory();
    const dispatch = useDispatch();

    useEffect( () => {
        if (success) {
            async function clearData() {
                await dispatch(clearUserData());
            }
            
            clearData();
            cookie.remove('auth');
            history.push(routerRoutes.home);
        }
        // eslint-disable-next-line
    }, [success])

    const initialValues = {
        password: initialFormValues.password,
        confirmPassword: initialFormValues.confirmPassword
    }

    const validationSchema = yup.object({
        password: yup
                .string()
                .required(formErrorMessages.required),
        confirmPassword: yup
                .string()
                .oneOf([yup.ref('password'), null], formErrorMessages.passwordMatch)
    });

    const handleDeleteAccount = async ({password}) => {
        if (!userId) {
            return;
        }

        setDeleteAccountStatus( prevStatus => {
            return {
                ...prevStatus,
                isDeleting : true
            }
        });

        const endpoint = `api/deleteAccount?id=${userId}`;
        const body = {
            password
        };

        try {
            const { data } = await axios.post(endpoint, body);
            const { error } = data; 

            if (error) {
                setDeleteAccountStatus({
                    isDeleting: false,
                    error,
                    success: false
                });

                return;
            }

            setDeleteAccountStatus({
                isDeleting: false,
                error: null,
                success: true
            });
        } catch (error) {
            console.error(error);

            setDeleteAccountStatus({
                isDeleting: false,
                error,
                success: false
            });
        }
    }

    const {
        values,
        errors,
        isValid,
        handleChange,
        handleBlur,
        handleSubmit,
        touched
    } = useFormik({
        initialValues,
        validationSchema,
        onSubmit: handleDeleteAccount
    })

    const controlEvents = {
        handleChange,
        handleBlur
    }

    return {
        values,
        errors,
        isValid,
        controlEvents,
        touched,
        handleSubmit,
        isDeleting,
        error
    }
}