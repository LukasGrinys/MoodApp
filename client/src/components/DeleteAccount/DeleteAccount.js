import React, { Fragment } from 'react';
import BackButton from '../Back/BackButton';
import { Wrapper, Heading, Label, Input, Footer, ErrorMessage } from '../Forms/Forms';
import ErrorBox from '../ErrorBox/ErrorBox';
import Button from '../Button/Button';
import { useDeleteAccount } from '../../hooks/DeleteAccount/useDeleteAccount';

const DeleteAccount = () => {
    const {
        values,
        errors,
        isValid,
        controlEvents,
        touched,
        handleSubmit,
        isDeleting,
        error
    } = useDeleteAccount();

    return (
        <Fragment>
            <BackButton/>
            <Wrapper>
                <ErrorBox>
                    <strong>Warning: </strong> deleting your account will 
                    permanently delete all your data, including all logs
                    stored in the database.
                </ErrorBox><br/>
                <Heading>Delete account:</Heading>
                <Label>Enter your passsword:</Label>
                <Input
                    name="password"
                    value={values.password}
                    type="password"
                    controlEvents={controlEvents}
                />
                {
                    touched.password && errors.password && (
                        <ErrorMessage>
                            {errors.password}
                        </ErrorMessage>
                    )
                }
                <Label>Confirm your password:</Label>
                <Input
                    name="confirmPassword"
                    value={values.confirmPassword}
                    type="password"
                    controlEvents={controlEvents}
                />
                {
                    touched.confirmPassword && errors.confirmPassword && (
                        <ErrorMessage>
                            {errors.confirmPassword}
                        </ErrorMessage>
                    )
                }
                <Footer>
                    <Button
                        disabled={!isValid || isDeleting}
                        handleClick={handleSubmit}
                    >
                        Delete account
                    </Button>
                </Footer>
                {
                    error && !isDeleting && (
                        <ErrorBox>
                            {error}
                        </ErrorBox>
                    ) 
                }
            </Wrapper>
        </Fragment>
    )
}

export default DeleteAccount