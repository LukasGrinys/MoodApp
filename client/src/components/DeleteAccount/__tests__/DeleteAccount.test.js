import React from 'react';
import { Provider } from 'react-redux';
import DeleteAccount from '../DeleteAccount';
import { mount } from 'enzyme';
import configureStore from 'redux-mock-store';
import { testUser } from '../../../../__mocks__/userMock';
import { formErrorMessages } from '../../../constants/formErrorMessages';

const mockStore = configureStore();

jest.mock('react-router-dom', () => {
    return {
        useHistory: jest.fn()
    }
});

const store = mockStore({
    user: {
        userData: testUser
    },
    logs: {}
});

const wrapper = mount(
    <Provider store={store}>
        <DeleteAccount/>
    </Provider>
);

describe('DeleteAccount page UI', () => {
    test('Renders properly', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('Has back button', () => {
        expect(wrapper.find('BackButton').length).toBe(1);
    });

    test('Has inputs for passwords', () => {
        expect(wrapper.find('input[name="password"]').length).toBe(1);
        expect(wrapper.find('input[name="confirmPassword"]').length).toBe(1);
    });

    test('Inputs have proper initial props', () => {
        expect(wrapper.find('input[name="password"]').props()).toEqual({
            className: 'textInput',
            name: 'password',
            onBlur: expect.any(Function),
            onChange: expect.any(Function),
            type: 'password',
            value: ''
        });

        expect(wrapper.find('input[name="confirmPassword"]').props()).toEqual({
            className: 'textInput',
            name: 'confirmPassword',
            onBlur: expect.any(Function),
            onChange: expect.any(Function),
            type: 'password',
            value: ''
        });
    });

    test('Has DeleteAccount button', () => {
        expect(wrapper.find('button').length).toBe(1);
        expect(wrapper.find('button').text()).toBe('Delete account');
    });

    test('DeleteAccount button has proper props', () => {
        expect(wrapper.find('button').props()).toEqual({
            className: 'button btnPrimary',
            disabled: false,
            onClick: expect.any(Function),
            type: 'button',
            children: 'Delete account'
        })
    });
});

describe('DeleteAccount page logic', () => {
    const testPassword = 'TestTest';
    const differentPassword = 'ThisIsDifferent'

    test('Input field value updates when typed in', () => {
        wrapper.find('input[name="password"]').simulate('change', {
            target: {
                name: 'password',
                value: testPassword
            }
        });
        expect(wrapper.find('input[name="password"]').prop('value')).toBe(testPassword);
    });

    test('Shows error message if passwords do not match', () => {
        wrapper.find('input[name="confirmPassword"]').simulate('change', {
            target: {
                name: 'confirmPassword',
                value: differentPassword
            }
        });

        wrapper.update();
        
        setTimeout( () => {
            expect(wrapper.find('ErrorBox').at(1).text()).toBe(formErrorMessages.passwordMatch);
            done();
        }, 0);
    });

    test('Button is disabled if values are invalid (passwords do not match)', () => {
        wrapper.find('input[name="password"]').simulate('change', {
            target: {
                name: 'password',
                value: testPassword
            }
        });
        wrapper.find('input[name="confirmPassword"]').simulate('change', {
            target: {
                name: 'confirmPassword',
                value: differentPassword
            }
        });
        wrapper.update();

        setTimeout( () => {
            expect(wrapper.find('button').prop('disabled')).toBeTruthy();
            done();
        }, 0);
    });

    test('Button is enabled if values are valid', () => {
        wrapper.find('input[name="password"]').simulate('change', {
            target: {
                name: 'password',
                value: testPassword
            }
        });
        wrapper.find('input[name="confirmPassword"]').simulate('change', {
            target: {
                name: 'confirmPassword',
                value: testPassword
            }
        });
        wrapper.update();

        setTimeout( () => {
            expect(wrapper.find('button').prop('disabled')).toBeFalsy();
            done();
        }, 0);
    });

    test('Button click calls handleDeleteAccount with the password', () => {
        const handleDeleteAccount = jest.fn();
        wrapper.find('button').simulate('click');

        setTimeout( () => {
            expect(handleDeleteAccount).toHaveBeenCalledTimes(1);
            expect(handleDeleteAccount).toHaveBeenCalledWith({
                password: testPassword
            });
            done();
        }, 0);
    });
})
