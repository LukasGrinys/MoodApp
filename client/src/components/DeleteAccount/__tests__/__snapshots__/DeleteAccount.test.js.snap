// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeleteAccount page UI Renders properly 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <DeleteAccount>
    <BackButton>
      <div
        className="container"
      >
        <div
          className="button"
          onClick={[Function]}
        >
          <FontAwesome
            name="arrow-left"
          >
            <span
              aria-hidden={true}
              className="fa fa-arrow-left"
            />
          </FontAwesome>
          <div
            className="text"
          >
            Go back
          </div>
        </div>
      </div>
    </BackButton>
    <Wrapper>
      <div
        className="formWrapper"
      >
        <ErrorBox>
          <div
            className="container"
          >
            <strong>
              Warning: 
            </strong>
             deleting your account will permanently delete all your data, including all logs stored in the database.
          </div>
        </ErrorBox>
        <br />
        <Heading>
          <h1
            className="heading"
          >
            Delete account:
          </h1>
        </Heading>
        <Label>
          <span
            className="label"
          >
            Enter your passsword:
          </span>
        </Label>
        <Input
          controlEvents={
            Object {
              "handleBlur": [Function],
              "handleChange": [Function],
            }
          }
          name="password"
          placeholder=""
          type="password"
          value=""
        >
          <input
            className="textInput"
            name="password"
            onBlur={[Function]}
            onChange={[Function]}
            type="password"
            value=""
          />
        </Input>
        <Label>
          <span
            className="label"
          >
            Confirm your password:
          </span>
        </Label>
        <Input
          controlEvents={
            Object {
              "handleBlur": [Function],
              "handleChange": [Function],
            }
          }
          name="confirmPassword"
          placeholder=""
          type="password"
          value=""
        >
          <input
            className="textInput"
            name="confirmPassword"
            onBlur={[Function]}
            onChange={[Function]}
            type="password"
            value=""
          />
        </Input>
        <Footer>
          <div
            className="formFooter"
          >
            <Button
              disabled={false}
              handleClick={[Function]}
            >
              <button
                className="button btnPrimary"
                disabled={false}
                onClick={[Function]}
                type="button"
              >
                Delete account
              </button>
            </Button>
          </div>
        </Footer>
      </div>
    </Wrapper>
  </DeleteAccount>
</Provider>
`;
